### PR TITLE
IMP in KORE

### LINKED ISSUE #70: KORE Translation: IMP

### LINKED ISSUE #71: KORE Translation: owise rules

### LINKED ISSUE #75: KORE Translation: as patterns

### LINKED ISSUE #79: KORE translation: comm, assoc, idem, unit axioms

### LINKED ISSUE #87: KORE Translation: functional/constructor attributes

### PR DESCRIPTION
**Pull request by dwightguth**
Adds code to automatically generate a kore definition from imp.k.

Fixes #70 
Fixes #71 
Fixes #75 
Fixes #79 
Fixes #87 

Generated imp.kore is [here](https://gist.github.com/dwightguth/d14a57ad9eded9f021b1077f55200344)

Ready for review.

### CODE REVIEW

**Comment by dwightguth** [comment]
Jenkins: test this please

File: k-distribution/include/builtin/domains.k [hunk 1]
```diff
@@ -287,8 +287,8 @@ module BOOL
   rule false andBool _:Bool => false
   rule _:Bool andBool false => false
 
-  rule true andThenBool K:K => K
-  rule K:K andThenBool true => K
+  rule true andThenBool K::Bool => K
+  rule K::Bool andThenBool true => K
```
Comment (radumereuta) [inline k-distribution/include/builtin/domains.k:291]: Is the sort annotation still needed after the syntax change?

File: java-backend/src/main/resources/org/kframework/backend/java/symbolic/hooks.properties [hunk 2]
```diff
@@ -157,7 +157,7 @@ STRING.floatFormat : org.kframework.backend.java.builtins.BuiltinStringOperation
 FLOAT.int2float : org.kframework.backend.java.builtins.BuiltinFloatOperations.int2float
 FLOAT.float2int : org.kframework.backend.java.builtins.BuiltinFloatOperations.float2int
 STRING.token2string : org.kframework.backend.java.builtins.BuiltinStringOperations.token2string
-STRING.string2token : org.kframework.backend.java.builtins.BuiltinStringOperations.string2token
+STRING.parseToken : org.kframework.backend.java.builtins.BuiltinStringOperations.string2token
```
Comment (radumereuta) [inline java-backend/src/main/resources/org/kframework/backend/java/symbolic/hooks.properties:160]: Should the builtin operator be renamed also? Or duplicated and keep `parseToken` as an alias? Not really worth looking into, just a thought.

File: k-distribution/include/builtin/domains.k [hunk 3]
```diff
@@ -590,19 +589,31 @@ module ID-PROGRAM-PARSING
               | #UpperId                                             [notInRules, token, autoReject]
```
Comment (radumereuta) [inline k-distribution/include/builtin/domains.k:589]: I think `notInRules` is no longer used.

File: k-distribution/include/builtin/domains.k [hunk 4]
```diff
@@ -590,19 +589,31 @@ module ID-PROGRAM-PARSING
...
 endmodule
 
-module ID
-  imports STRING
+module ID-SYNTAX
+  imports INT-SYNTAX
+  imports STRING-SYNTAX
 
-  syntax Id
+  syntax Id [token]
```
Comment (radumereuta) [inline k-distribution/include/builtin/domains.k:596]: I'm not up to date with the latest changes in K5. Is the meaning of `token` changed or extended?

File: k-distribution/include/builtin/domains.k [hunk 5]
```diff
@@ -590,19 +589,31 @@ module ID-PROGRAM-PARSING
...
 
   syntax String ::= Id2String ( Id )    [function, hook(STRING.token2string)]
-  syntax Id ::= String2Id ( String )    [function]
-  rule String2Id(S:String) => #parseToken("Id", S)
-
+  syntax Id ::= String2Id (String) [function, hook(STRING.string2token)]
```
Comment (radumereuta) [inline k-distribution/include/builtin/domains.k:599]: I see in the list of hooks there is a `string2id`, why use `string2token` here?
Wait, you just removed `string2token`. Can you take another look at this?

File: kernel/src/main/java/org/kframework/compile/GenerateSentencesFromConfigDecl.java [hunk 6]
```diff
@@ -221,13 +221,15 @@ private static K getLeafInitializer(K leafContents) {
             @Override
             public K apply(KToken k) {
                 if (k.sort().equals(Sorts.KConfigVar())) {
-                    return KApply(KLabel("Map:lookup"), KVariable("Init"), k);
+                    return KApply(KLabel("Map:lookup"), INIT, k);
```
Comment (radumereuta) [inline kernel/src/main/java/org/kframework/compile/GenerateSentencesFromConfigDecl.java:224]: Maybe extract `Map:lookup` as a constant somewhere also?

**Review by radumereuta — APPROVED** [review]
I'm not up to date with K5, so I only added a few questions and remarks.

**Review by traiansf — APPROVED** [review]
I'm satisfied with the current result. Of course, once we sort out pending issues we might need to slightly adapt this, but for now it should be enough to help us move forward.

File: k-distribution/include/builtin/domains.k (hunk 4 above)
Comment (dwightguth) [reply k-distribution/include/builtin/domains.k:596]: This is being used to indicate that the abstract syntax of a sort contains KToken, even though it's not part of the parser.

File: k-distribution/include/builtin/domains.k (hunk 5 above)
Comment (dwightguth) [reply k-distribution/include/builtin/domains.k:599]: This is the version of the module used by the ocaml and kore backends, not by the java backend

**Pull request merged**